packages/react-native-audio-api/android/src/main/cpp/audioapi/android/core/utils/AndroidRecorderCallback.cpp

```diff
@@ -0,0 +1,187 @@
+  processingBuffer_ = ma_malloc(
+      processingBufferLength_ * channelCount_ *
+          ma_get_bytes_per_sample(ma_format_f32),
+      NULL);
+}
+
+void AndroidRecorderCallback::cleanup() {
+  if (converter_ != nullptr) {
+    ma_data_converter_uninit(converter_.get(), NULL);
+    converter_.reset();
+  }
+
+  if (processingBuffer_ != nullptr) {
+    ma_free(processingBuffer_, NULL);
+    processingBuffer_ = nullptr;
+    processingBufferLength_ = 0;
+  }
+
+  for (size_t i = 0; i < circularBus_.size(); ++i) {
+    circularBus_[i]->zero();
+  }
+}
+
+void AndroidRecorderCallback::receiveAudioData(void *data, int numFrames) {
+  ma_uint64 inputFrameCount = numFrames;
+  ma_uint64 outputFrameCount = 0;
+
+  if ((float)streamSampleRate_ == sampleRate_ &&
+      streamChannelCount_ == channelCount_) {
+    deinterleaveAndWriteAudioData(data, numFrames);
+    emitAudioData();
+    return;
+  }
+
+  ma_data_converter_get_expected_output_frame_count(
+      converter_.get(), inputFrameCount, &outputFrameCount);
+
+  ma_data_converter_process_pcm_frames(
+      converter_.get(),
+      data,
+      &inputFrameCount,
+      processingBuffer_,
+      &outputFrameCount);
+
+  deinterleaveAndWriteAudioData(
+      processingBuffer_, static_cast<int>(outputFrameCount));
+  emitAudioData();
+}
+
+void AndroidRecorderCallback::deinterleaveAndWriteAudioData(
+    void *data,
+    int numFrames) {
+  auto *inputData = static_cast<float *>(data);
+
+  for (size_t channel = 0; channel < channelCount_; ++channel) {
+    float *channelData = deinterleavingArray_->getData();
+
+    for (int frame = 0; frame < numFrames; ++frame) {
+      channelData[frame] = inputData[frame * channelCount_ + channel];
+    }
+
+    circularBus_[channel]->push_back(channelData, numFrames);
+  }
+}
+
+void AndroidRecorderCallback::emitAudioData() {
+  while (circularBus_[0]->getNumberOfAvailableFrames() >= bufferLength_) {
+    auto bus =
+        std::make_shared<AudioBus>(bufferLength_, channelCount_, sampleRate_);
+
+    for (int i = 0; i < channelCount_; ++i) {
+      auto *outputChannel = bus->getChannel(i)->getData();
+      circularBus_[i]->pop_front(outputChannel, bufferLength_);
+    }
+
+    invokeCallback(bus, (int)bufferLength_);
+  }
+}
+
+void AndroidRecorderCallback::invokeCallback(
+    const std::shared_ptr<AudioBus> &bus,
+    int numFrames) {
+  auto audioBuffer = std::make_shared<AudioBuffer>(bus);
+  auto audioBufferHostObject =
+      std::make_shared<AudioBufferHostObject>(audioBuffer);
+
+  std::unordered_map<std::string, EventValue> eventPayload = {};
+  eventPayload.insert({"buffer", audioBufferHostObject});
+  eventPayload.insert({"numFrames", numFrames});
+
+  if (audioEventHandlerRegistry_) {
+    audioEventHandlerRegistry_->invokeHandlerWithEventBody(
+        "audioReady", callbackId_, eventPayload);
+  }
+}
+
+void AndroidRecorderCallback::sendRemainingData() {
+  auto numberOfFrames = circularBus_[0]->getNumberOfAvailableFrames();
```

Comment:
ideally circular bus should be a separate class instead of plain vector of circular arrays